### PR TITLE
Proxyclient ASsyncTCP + secure connetions in ASyncTCP

### PR DESCRIPTION
**Pull request by rkluwen**
This pull request implements the CProxyClient class as a base class of ASyncTCP.
At the same time the ASyncTCP class has been extended to be able to connect via ssl sockets.

This patch hasn't been tested for a long period of time, so a message to all testers is to use this branch for a while.

### CODE REVIEW

**Comment by rkluwen** [comment]
At the same time a lot of shared pointers are removed in favor of non-pointer members.

File: hardware/ASyncTCP.h [hunk 1]
```diff
@@ -1,11 +1,15 @@
 #pragma once
 
 #include <stddef.h>                        // for size_t
+#include <queue>						   // for write queue
 #include <boost/asio/deadline_timer.hpp>   // for deadline_timer
 #include <boost/asio/io_service.hpp>       // for io_service
 #include <boost/asio/ip/tcp.hpp>           // for tcp, tcp::endpoint, tcp::s...
+#include <boost/asio/ssl.hpp>			   // for secure sockets
```
Comment (gizmocuz) [inline hardware/ASyncTCP.h:8]: I think you have to use the preprocessor 'WWW_ENABLE_SSL' also in this class (openssl has to be found correct?, see webserver/server.hpp)

File: hardware/ASyncTCP.h [hunk 2]
```diff
@@ -15,7 +19,7 @@ namespace boost { namespace system { class error_code; } }
 class ASyncTCP
 {
 protected:
-	ASyncTCP();
+	ASyncTCP(bool secure = false);
```
Comment (gizmocuz) [inline hardware/ASyncTCP.h:22]: could you make the argument 'const' ?

File: hardware/ASyncTCP.cpp [hunk 3]
```diff
@@ -260,6 +310,18 @@ void ASyncTCP::write_end(const boost::system::error_code& error)
 				StartReconnect();
 			}
 		}
+		else {
+			boost::unique_lock<boost::mutex> lock(writeMutex);
+			if (writeQ.size() > 0) {
```
Comment (gizmocuz) [inline hardware/ASyncTCP.cpp:315]: Should this be a 'while > 0' ? could it be that there are more items in writeQ ?
Please use if (!writeQ.empty())

File: hardware/ASyncTCP.h [hunk 4]
```diff
@@ -83,6 +90,11 @@ class ASyncTCP
 	std::shared_ptr<std::thread> 	m_tcpthread;
 	boost::asio::io_service::work 	m_tcpwork; // Create some work to keep IO Service alive
 
+	boost::asio::ssl::context		mContext; // ssl context
 	boost::asio::ip::tcp::socket	mSocket;
+	boost::asio::ssl::stream<boost::asio::ip::tcp::socket> mSslSocket; // the ssl socket
 	boost::asio::ip::tcp::endpoint	mEndPoint;
+	std::queue<std::string>			writeQ; // we need a write queue to allow concurrent writes
```
Comment (gizmocuz) [inline hardware/ASyncTCP.h:97]: would a dequeue be better/faster?

File: webserver/proxyclient.h [hunk 5]
```diff
@@ -23,45 +24,27 @@ namespace http {
...
-				size_t bytes_transferred);
-
-			void ReadMore();
-
 			void WebsocketGetRequest();
 			bool parse_response(const char *data, size_t len);
 			std::string websocket_key;
 			std::string compute_accept_header(const std::string &websocket_key);
-			std::string SockWriteBuf;
+			std::string readbuf;
```
Comment (gizmocuz) [inline webserver/proxyclient.h:45]: If possible, please start 'all' member variables with m_

**Review by gizmocuz — COMMENTED** [review]
Thanks for the PR, some small questions/suggestions

**Comment by rkluwen** [comment]
Should this be a 'while > 0' ? could it be that there are more items in writeQ ?

- No, there should be only one outstanding async_write(). That's exactly the purpose of writeQ.

would a dequeue be better/faster?

- I'm not sure

About the other comments: Roger & Wilco. I will implement those and come back with an update on this PR.

**Comment by rkluwen** [comment]
The pull request doesn't work on the Pi.

File: hardware/ASyncTCP.h (hunk 1 above)
Comment (rkluwen) [reply hardware/ASyncTCP.h:8]: Doine

File: hardware/ASyncTCP.h (hunk 2 above)
Comment (rkluwen) [reply hardware/ASyncTCP.h:22]: Done

File: hardware/ASyncTCP.cpp (hunk 3 above)
Comment (rkluwen) [reply hardware/ASyncTCP.cpp:315]: writeQ,empty() used now. The other comment was a negative.

File: hardware/ASyncTCP.h (hunk 4 above)
Comment (rkluwen) [reply hardware/ASyncTCP.h:97]: Yeah... Makes much more sense. Done.

File: webserver/proxyclient.h (hunk 5 above)
Comment (rkluwen) [reply webserver/proxyclient.h:45]: Done.

**Pull request closed without merging**